File: tasks/native-runtime/browser_controller_parity_probe.py
"""browser controller/engine-state parity probe, Python side.

Pins `session.controller_kind` and the controller a *result* reports across the
states the oracle distinguishes: a session nothing has answered yet, a refused
action, a dispatched action, and a failed one. Also pins the engine-availability
answer that selects the controller.

**The precondition this runs under.** The native runtime has no Playwright
engine, so the deployment it corresponds to is the one where
`playwright_available()` is false and the static fallback is the controller. The
`playwright` module is therefore made unimportable before anything imports it.
That is the documented environment of the static fallback, not a stand-in for
the code under test: the controller-selection state machine runs untouched.

The host's real answer is reported on **stderr** and stays outside the compared
bytes on purpose — the two sides are *supposed* to disagree there, and that
disagreement is exactly the engine gap this probe does not cover.

Usage::

    python tasks/native-runtime/browser_controller_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example browser_controller_parity_probe > ../rust.json
"""

from __future__ import annotations

import importlib.util
import json
import sys
from pathlib import Path
from typing import Any, cast

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

HOST_PLAYWRIGHT_IMPORTABLE = importlib.util.find_spec("playwright") is not None

# The precondition: no browser engine is importable, so `playwright_available()`
# answers false and `_create_controller` takes the static fallback. A `None` entry
# makes `import playwright.sync_api` raise, which is what the oracle probes for.
sys.modules["playwright"] = cast(Any, None)

from deepseek_infra.core import config  # noqa: E402
from deepseek_infra.infra.browser import actions, session  # noqa: E402
from deepseek_infra.infra.browser.controller import playwright_available  # noqa: E402

PRIVATE_URL = "http://127.0.0.1:8000/private"


def main() -> int:
    config.BROWSER_CONTROL_ENABLED = True
    config.BROWSER_REQUIRE_CONFIRM = True
    config.BROWSER_ALLOW_PRIVATE_HOSTS = False
    config.BROWSER_HEADLESS = True
    session.reset_sessions_for_tests()

    fixture = REPO / "tests" / "fixtures" / "browser" / "basic.html"
    if not fixture.exists():
        raise SystemExit(f"missing fixture: {fixture}")
    uri = "file:///" + fixture.as_posix()

    # 1. A session exists but nothing has answered: no controller is recorded yet.
    fresh = actions.execute_browser_action({"action": ""})

    # 2. Refused on the URL, which happens *before* a controller is built.
    blocked = actions.execute_browser_action({"action": "open_url", "url": PRIVATE_URL})

    # 3. Dispatched: a controller answered, so the session records its kind.
    opened = actions.execute_browser_action({"action": "open_url", "url": uri})
    session_id = str(opened["session"]["browserSessionId"])

    # 4. Failed. `download` with no target raises from the controller.
    try:
        actions.execute_browser_action({"action": "download", "sessionId": session_id, "confirmed": True})
        raised = False
    except Exception:
        raised = True

    # A refused action reads the stored session back without touching it, so it is
    # how the failed state is observable through the public API.
    observed = actions.execute_browser_action({"action": "open_url", "sessionId": session_id, "url": PRIVATE_URL})

    # 5. A later success reports the live controller, but the recorded kind is
    #    sticky — `controller_for` hands back the cached controller without
    #    touching the session again. `extract_links` is used because `read_page`
    #    also writes a snapshot into the Python-owned media store.
    read = actions.execute_browser_action({"action": "extract_links", "sessionId": session_id})

    def text(payload: dict[str, Any], key: str) -> str:
        return str(payload.get(key) or "")

    def prefixed(value: Any) -> bool:
        return str(value or "").startswith("failed:")

    out: dict[str, Any] = {
        "blocked-code": text(blocked, "code"),
        "blocked-controller": text(blocked["session"], "controller"),
        "blocked-risk": text(blocked["safety"], "risk"),
        "dispatched-result-controller": text(opened["result"], "controller"),
        "dispatched-session-controller": text(opened["session"], "controller"),
        "dispatched-session-status": text(opened["session"], "status"),
        "engine": text(opened["session"], "engine"),
        "engine-available": bool(playwright_available()),
        "failed-action-raised": raised,
        "failed-controller-is-prefixed": prefixed(observed["session"]["controller"]),
        "failed-status": text(observed["session"], "status"),
        "fresh-controller": text(fresh["session"], "controller"),
        "fresh-status": text(fresh["session"], "status"),
        "live-controller-after-failure": text(read["result"], "controller"),
        "sticky-controller-is-prefixed": prefixed(read["session"]["controller"]),
    }
    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")

    # Diagnostics, deliberately outside the compared bytes.
    print(f"# compared under: playwright importable = {HOST_PLAYWRIGHT_IMPORTABLE} (hidden for the run)", file=sys.stderr)
    print(f"# host truth: engine-available would be {HOST_PLAYWRIGHT_IMPORTABLE}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/browser_controller_parity_probe.rs
//! browser controller/engine-state parity probe, Rust side.
//!
//! Pins `session.controller_kind` and the controller a *result* reports across
//! the states the oracle distinguishes: a session nothing has answered yet, a
//! refused action, a dispatched action, and a failed one. Also pins the
//! engine-availability answer that selects the controller.
//!
//! Run against `tasks/native-runtime/browser_controller_parity_probe.py`; the
//! two outputs must be byte-identical.
//!
//! Usage:
//!
//! ```text
//! python tasks/native-runtime/browser_controller_parity_probe.py > python.json
//! cd rust && cargo run -p deepseek-policy --example browser_controller_parity_probe > ../rust.json
//! ```
//!
//! Keys are emitted from a `BTreeMap` so the order is sorted regardless of
//! whether `serde_json` is built with `preserve_order` (which the workspace
//! turns on for some crates). The Python side uses `sort_keys=True` for the same
//! bytes.

use std::collections::BTreeMap;
use std::path::PathBuf;

use deepseek_policy::browser::{
    ENGINE_PLAYWRIGHT, execute_browser_action, playwright_available, reset_sessions_for_tests,
};
use deepseek_policy::browser_safety::BrowserSettings;
use deepseek_policy::entropy::SystemEntropy;
use serde_json::{Value, json};

const PRIVATE_URL: &str = "http://127.0.0.1:8000/private";

fn fixture() -> (String, BrowserSettings) {
    let fixture = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../../tests/fixtures/browser/basic.html");
    assert!(fixture.exists(), "missing {}", fixture.display());
    let display = fixture
        .canonicalize()
        .unwrap_or(fixture.clone())
        .to_string_lossy()
        .replacen(r"\\?\", "", 1);
    let uri = format!("file:///{}", display.replace('\\', "/"));
    let settings = BrowserSettings {
        enabled: true,
        require_confirm: true,
        allow_private_hosts: false,
        headless: true,
        fixture_roots: vec![fixture.parent().expect("fixture dir").to_path_buf()],
    };
    (uri, settings)
}

fn text(value: &Value, key: &str) -> Value {
    json!(value.get(key).and_then(Value::as_str).unwrap_or(""))
}

fn prefixed(value: &Value) -> bool {
    value
        .as_str()
        .is_some_and(|found| found.starts_with("failed:"))
}

fn main() {
    reset_sessions_for_tests();
    let (uri, settings) = fixture();
    let entropy = SystemEntropy;
    let browser = |payload: Value| execute_browser_action(&payload, &settings, &entropy);
    let mut out: BTreeMap<String, Value> = BTreeMap::new();

    // 1. A session exists but nothing has answered: no controller is recorded yet.
    let fresh = browser(json!({"action": ""})).expect("empty action returns a payload");

    // 2. Refused on the URL, which the oracle does *before* it builds a controller.
    let blocked = browser(json!({"action": "open_url", "url": PRIVATE_URL}))
        .expect("refused action returns a payload");

    // 3. Dispatched: a controller answered, so the session records its kind.
    let opened = browser(json!({"action": "open_url", "url": uri})).expect("fixture opens");
    let session_id = opened["session"]["browserSessionId"]
        .as_str()
        .expect("session id")
        .to_string();

    // 4. Failed: `download` is not ported, so the dispatch refuses after the gate.
    let failed = browser(json!({"action": "download", "sessionId": session_id, "confirmed": true}));
    let raised = failed.is_err();

    // A refused action reads the stored session back without touching it, so it is
    // how the failed state is observable through the public API.
    let observed = browser(json!({
        "action": "open_url",
        "sessionId": session_id,
        "url": PRIVATE_URL,
    }))
    .expect("refused action returns a payload");

    // 5. A later success reports the live controller, but the recorded kind is
    //    sticky — the oracle hands back its cached controller without touching
    //    the session again.
    let read = browser(json!({"action": "read_page", "sessionId": session_id}))
        .expect("read_page on a live session");

    out.insert("blocked-code".to_string(), text(&blocked, "code"));
    out.insert(
        "blocked-controller".to_string(),
        text(&blocked["session"], "controller"),
    );
    out.insert("blocked-risk".to_string(), text(&blocked["safety"], "risk"));
    out.insert(
        "dispatched-result-controller".to_string(),
        text(&opened["result"], "controller"),
    );
    out.insert(
        "dispatched-session-controller".to_string(),
        text(&opened["session"], "controller"),
    );
    out.insert(
        "dispatched-session-status".to_string(),
        text(&opened["session"], "status"),
    );
    out.insert("engine".to_string(), json!(ENGINE_PLAYWRIGHT));
    out.insert(
        "engine-available".to_string(),
        json!(playwright_available()),
    );
    out.insert("failed-action-raised".to_string(), json!(raised));
    out.insert(
        "failed-controller-is-prefixed".to_string(),
        json!(prefixed(&observed["session"]["controller"])),
    );
    out.insert(
        "failed-status".to_string(),
        text(&observed["session"], "status"),
    );
    out.insert(
        "fresh-controller".to_string(),
        text(&fresh["session"], "controller"),
    );
    out.insert(
        "fresh-status".to_string(),
        text(&fresh["session"], "status"),
    );
    out.insert(
        "live-controller-after-failure".to_string(),
        text(&read["result"], "controller"),
    );
    out.insert(
        "sticky-controller-is-prefixed".to_string(),
        json!(prefixed(&read["session"]["controller"])),
    );

    let mut encoded = serde_json::to_string_pretty(&out).expect("serialize");
    encoded.push('\n');
    print!("{encoded}");
}
